### ISSUE TITLE
Crash: `You must feed a value for placeholder tensor 'z' with dtype float`

### ISSUE THREAD

**Issue by gnperdue** [hep-lbdl/CaloGAN#1]
Hello,

First, thanks for the very nice repository and documentation. I've been trying to run the `train.py` module and running into trouble. 

You can try to replicate my error by doing the following...

This will set up my software environment:
* `git clone https://github.com/gnperdue/BashDotFiles.git`
* `cd BashDotFiles/anaconda_configs && bash miniconda_py2dl.sh`
* `export PATH=$HOME/miniconda2/bin && source activate py2dl`

I have a version checker there that shows what this pulls down for this particular setup:
```
$ python pyverchecker.py
Python version: 2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:05:08)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
pandas version: 0.20.1
matplotlib version: 2.0.2
numpy version: 1.12.1
scipy version: 0.19.0
IPython version: 5.3.0
sklearn version: 0.18.1
skimage version: 0.13.0
Missing protobuf
Missing numexpr
Missing sympy
tensorflow version: 1.0.0
pymc3 version: 3.0
theano version: 0.9.0
h5py version: 2.7.0
Using TensorFlow backend.
keras version: 2.0.2
Possible problem with module: xlrd, 'module' object has no attribute '__version__'
Missing lasagne
yaml version: 3.12
Missing nltk
```

Next, I retrieved the data you provided and updated  the `models/particles.yaml` file:

```
# put the paths to actual data here!
positron: 'data/eplus.hdf5'
# pion: 'data/piplus.hdf5'
# gamma: 'data/gamma.hdf5'
```

I also opened the file and poked around and everything looks fine. My Keras configuration is like so:

```
.keras$ more keras.json
{
    "epsilon": 1e-07,
    "floatx": "float32",
    "image_data_format": "channels_last",
    "backend": "tensorflow"
}
```

`python -m models.train -h` produces sensible results. However, `python -m models.train models/particles.yaml` produces output like:

```
(py2dl) CaloGAN$ python -m models.train models/particles.yaml
Using TensorFlow backend.
2017-05-30 17:32:16,289 - models.train[INFO]: 1 particle types found.
2017-05-30 17:32:16,526 - models.train[INFO]: Building discriminator
...
Caused by op u'z', defined at:
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/perdue/Dropbox/ArtificialIntelligence/DSHEP/CaloGAN/models/train.py", line 302, in <module>
    latent = Input(shape=(latent_size, ), name='z')
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/keras/engine/topology.py", line 1388, in Input
    input_tensor=tensor)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/keras/engine/topology.py", line 1299, in __init__
    name=self.name)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 349, in placeholder
    x = tf.placeholder(dtype, shape=shape, name=name)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 1520, in placeholder
    name=name)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2149, in _placeholder
    name=name)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 763, in apply_op
    op_def=op_def)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2395, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/Users/perdue/miniconda2/envs/py2dl/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1264, in __init__
    self._traceback = _extract_stack()

InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'z' with dtype float
[[Node: z = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
```

I walked through the code in the debugger for a while and could not figure out what was causing the crash. It isn't clear to me how the `latent` input (named `'z'`, and part of `generator_inputs` in the `generator` network definition) enters where it does, as the crash is here, in  `discriminator.train_on_batch`:

```
426
427             real_batch_loss = discriminator.train_on_batch(
428                 [image_batch_1, image_batch_2, image_batch_3, energy_batch],
429                 disc_outputs_real,
430                 loss_weights
431             )
432
```

Any thoughts or suggestions would be deeply appreciated. Thanks!

pax
Gabe

**Comment by mickypaganini** [comment]
That's a problem we have run into before. As far as I remember, it was solved in two ways, one being more reliable, the other being a total hack. Not to point fingers, but I think this is a problem with TensorFlow and how it builds the graph. In fact, we never had this issue with Theano (though you could try and confirm that that's indeed the case).
Solution 1: can you try upgrading to the latest TF version?
Solution 2: you would have to try to compile an intermediate model, and pay the graph gets built correctly --> if solution 1 doesn't work, we can probably try to help you do this. 

Long story short, it's a pain, but updating TF solved it for us.

**Comment by gnperdue** [comment]
I'll try to remember to test with Theano also. When using the TF backend though, this problem does indeed go away when using version 1.1. I don't see the crash anymore if I use this stack:

```
$ python pyverchecker.py
Python version: 2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:05:08)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
pandas version: 0.20.1
matplotlib version: 2.0.2
numpy version: 1.12.1
scipy version: 0.19.0
IPython version: 5.3.0
sklearn version: 0.18.1
skimage version: 0.13.0
tensorflow version: 1.1.0
theano version: 0.9.0
h5py version: 2.7.0
keras version: 2.0.4
yaml version: 3.12
```

Thanks!
